Ieee802154 changes required for sub-ghz support

tbursztyka opened the pull request.

Commit:
net/ieee802154: Add Sub-Ghz dedicated function on the radio API

Sub-Ghz bands have different limit of channels. 10, or more than a
thousand is actually possible. Thus the device needs to expose such
limit to the L2 which is unaware of frequency band logic. L2 will
then allow user to select a proper channel.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Verify in L2 the Sub-Ghz channel value

Depending on device's band, the upper channel limit can vary a lot in
Sub-Ghz. Thus verifying it directly in L2 before requesting it to the
device.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
hum, how can I make this PR dependent over https://github.com/zephyrproject-rtos/zephyr/pull/4094 so it won't show the commits of PR 4049?

jukkar commented on include/net/ieee802154_radio.h:
```diff
@@ -30,6 +30,7 @@ enum ieee802154_hw_caps {
 	IEEE802154_HW_FILTER		= BIT(2),
```
Something does not sound right in commit message, could it be written something like this instead:

```
Sub-Ghz bands can have different limit of channels, typically 10 or more than thousand are possible.
This means that the device needs to expose such limit to the L2 which is unaware of frequency band logic. L2 will then allow user to select a proper channel.
```

tbursztyka replied:
done

jukkar reviewed (changes requested):
Looks good, minor nit pick.

pfalcon commented:
@tbursztyka :

> hum, how can I make this PR dependent over #4094 so it won't show the commits of PR 4049?

You can't. PRs posted should be independent from each other and not include unrelated changes. Locally, a developer can have "integration branch" where they merge separate PR branches for testing/further local development. Yeah, like those Linux subsystem maintainers.

That may be not too flexible, but promotes a workflow where a succession of small, easy to review PRs posted, instead of one big, long to review one (or so I think that it does ;-) ).

pfalcon commented:
Commit: "net/ieee802154: Add lqi & rssi into net_pkt metadata structure":

Is it good model to store such data? For example, are there many users of such info which warrants spending memory on it for each and every packet? Secondly, RSSI/LQI apparently apply to a 15.4 packet, which for net_pkt becomes just a data fragment, and there can be many such data fragments with different RSSI/LQI's, how net_pkt's value get calculated? (Probably worth mention that in the commit message.)

pfalcon reviewed (changes requested):
The commit message for "net/ieee802154: Add lqi & rssi into net_pkt metadata structure" has:

> It's simpler and make the values relevantly tighten to the received packet.

I guess "tighten" is a wrong word here, it should be "tied".

dleach02 commented on subsys/net/ip/l2/ieee802154/ieee802154_utils.h:
```diff
@@ -68,4 +68,19 @@ static inline void ieee802154_filter_pan_id(struct net_if *iface,
+#ifdef CONFIG_NET_L2_IEEE802154_SUB_GHZ
+	const struct ieee802154_radio_api *radio = dev->driver_api;
+
+	if (radio->get_capabilities(dev) & IEEE802154_HW_SUB_GHZ) {
```
Why do we only verify SUB_GHZ channels and not the 2.4 stuff here?

dleach02 commented on subsys/net/ip/l2/ieee802154/Kconfig:
```diff
@@ -18,6 +18,10 @@ config NET_L2_IEEE802154_MGMT
 
+config NET_L2_IEEE802154_SUB_GHZ
+	bool
+	default n
```
This needs to have a "help" block to give the future user some guidance.

tbursztyka replied:
done

Commit:
net/ieee802154: Do not set the device UP by default

It will be up to the user to configure a valid channel, through
net_mgmt, and call net_if_up() in order to get the device up.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/lib: 15.4 interface is set UP after channel is configured

Once it has - at least - set up the channel, it can request the iface to
be put up.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Introduce helpers to call radio api function

Basically to make code nicer and more readable.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Use helpers to call radio API functions

Now the code is a bit cleaner.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Normalize parameter on radio API helper

For some reason, ieee802154_verify_channel had a device parameter
instead of an iface like all the others.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@pfalcon could you re-review? actually you reviewed patches that were not part of this PR.

pfalcon commented:
@tbursztyka : Yep, my concern was a typo in a commit message which could lead to confusion. Dismissed my -1. Can't say much about specific technical 15.4 changes, leaving it to other folks.

tbursztyka commented:
merged via CC1200 PR
